### ISSUE TITLE
Multi lingual annotations

### ISSUE THREAD

**Issue by stueynz** [json-schema-org/json-schema-vocabularies#10]
Historically OpenAPI and JSON Schema have worked together to document existing APIs. In a _contract first_ environment they are being asked to do more than merely document existing APIs; they are _defining_ APIs that are yet to be built.  JSON Schema is no longer just about payload validation, the annotations in a JSON Schema complete the definition of the payloads sufficient to define APIs that are yet to be built.

When defining enumeration values, JSON Schema lets us give each enumeration value annotations:

```json
  type: object
  properties:
    Name:
      type: string
    Gender:
       type: string
       oneOf: 
         - const: "M"
           title: "Male"
         - const: "F"
           title: "Female"
         - const: "D"
           title: "Diverse"
           description: "Used where gender known, and is non-binary"
 ```

When JSON Schema is used to _define_ APIs and message payloads, it is not unreasonable for software developers to choose to populate user interface elements (radio buttons, drop down lists, etc.) from the annotation values presented in the JSON schema.  In a multi-lingual environment, we may need multiple annotations, keyed by language.   

Expanding the definition of the `title` (and perhaps `description`) annotations to be either a single string or an object with properties keyed on language tags would be a good idea:

```json
  type: object
  properties:
    Name:
      type: string
    Gender:
       type: string
       oneOf: 
         - const: "M"
           title: {"en-NZ": "Male", "mi-NZ": "Tāne" }
         - const: "F"
           title: {"en-NZ": "Female", "mi-NZ": "Wahine" }
         - const: "D"
           title: {"en-NZ": "Diverse", "mi-NZ": "Ira tāngata kōwhiri kore" }
           description: "Used where gender known, and is non-binary"
 ```

**Comment by johandorland** [comment]
The example you gave when an interface is generated from a JSON schema is quite application specific. If it wants to include multilingual support in a schema I think the proper solution would be for it to define it's own vocabulary or work around it in other ways. Putting stuff like this in core only bloats the spec and puts unnecessary burdens on implementations.

**Comment by philsturgeon** [comment]
@brettz9 suggested the following in https://github.com/json-schema-org/json-schema-spec/issues/114:

> (Filing as a separate issue as per [#53 (comment)](https://github.com/json-schema-org/json-schema-spec/issues/53#issuecomment-248211146) )
> 
> As with HTML's `lang/xml:lang` properties, it would be useful to indicate the content language of a particular field in a standard manner. This might be used for proper font display (as in CJK languages) or for selectively showing content to users based on their locale.
> 
> I think a name "contentLang" for the property would avoid confusion at (falsely) thinking that this was necessarily indicating the language of the "title" or "description" itself.
> 
> Although it is probably not a feature that would be used for validation (since looking at code points might not be reliable or valid such such detection), `contentLang` does describe the data, placing constraints on how it is to be understood which brings it more into the world of schemas (unlike i18n of the schema itself).
> 
> Note that JSON-LD would not solve this unless one uses JSON-LD in the instance documents (which would prevent the benefit of having such a pseudo-constraint at the schema level).

Let's see if we can solve i18n as a vocab over here.

**Comment by brettz9** [comment]
I'd also like to offer a related suggestion if it might fit here---`format` set to a standard language code ([BCP 47](https://tools.ietf.org/html/bcp47)).

**Comment by handrews** [comment]
@brettz9 nothing should ever be added to `format` again.  Start a new set of keywords, don't further burden the most broken one that we haven't been able to nuke yet.

**Comment by brettz9** [comment]
Sorry, have some "brain fog" problems, so can make some dumb statements while iterating on my work.

So, it is understandable for not creating new official format values.

I have some further questions I'll post below, knowing this is not the right place for further discussion, but if I could kindly request your assistance in pointing me to an existing issue where such further discussion would be suitable.

----
But as far as starting a new set of keywords, for those who need to tack on something quickly for their project's needs and don't have the time to work on specifications (or to wait on them to be approved), I think people need a space to add multiple keywords without fear of namespace conflict.

Per your comment at https://github.com/json-schema-org/json-schema-vocabularies/issues/33#issuecomment-539267726 , "The bar for a new core keyword ($-prefix) is very high", suggesting that it is still possible that new core `$` may be used in the future, so it seems unsafe to use `$` alone as a namespace protection.

I found [this comment](https://github.com/json-schema-org/json-schema-vocabularies/issues/33#issuecomment-698561177) on adding a JSON Schema equivalent to a meta tag interesting, if this might be a kind of open-ended mechanism for adding custom keywords (or whatever may do so).

I really would hope that this or something like it could be prioritized so that one can begin to have multiple formats, and maybe to assist deprecating `format` as I don't believe the spec makes clear this is expected to happen.

This might even help with vocabulary development if experiments could be turned into standards. The meta approach I think has another aspect worthy of emulation too: having a registry open to anyone, like an equivalent to https://wiki.whatwg.org/wiki/MetaExtensions which could avoid namespace conflicts even for non-community adopted standards, and which allows for semi-formal as well as formal specifications.

**Comment by ThorFjelldalen** [comment]
Hi! Are there any new developments regarding internationalization of JSON Schema Validation, or minimal / best-practice workaround?

**Comment by gregsdennis** [comment]
@ThorFjelldalen in terms of official localization support, no.

However, with 2020-12, you will get any undefined keywords as annotations.  This means that you can include `title*` keywords for each language that you want to support.  Since `title` results in an annotation anyway, this change in processing the results shouldn't be too difficult.

Your schema from above:

```yaml
Gender:
  type: string
  oneOf: 
    - const: "M"
      title*: {"en-NZ": "Male", "mi-NZ": "Tāne" }
    - const: "F"
      title*: {"en-NZ": "Female", "mi-NZ": "Wahine" }
    - const: "D"
      title*: {"en-NZ": "Diverse", "mi-NZ": "Ira tāngata kōwhiri kore" }
      description: "Used where gender known, and is non-binary"
```

could become

```yaml
Gender:
  type: string
  oneOf: 
    - const: "M"
      title: "Male"
      title@mi-NZ: Tāne
    - const: "F"
      title: "Female"
      title@mi-NZ: Wahine
    - const: "D"
      title: "Diverse"
      title@mi-NZ: Ira tāngata kōwhiri kore
      description: "Used where gender known, and is non-binary"
```

Here, I'm using `title@<locale>` for the keys.  When evaluating an instance, e.g. `{"Gender": "F"}`, you'll get the result (in `Verbose` format):

```json
{
  "valid": true,
  "keywordLocation": "",
  "absoluteKeywordLocation": "https://json-everything.net/438509aca5",
  "instanceLocation": "",
  "annotations": [
    {
      "valid": false,
      "keywordLocation": "/oneOf/0",
      "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/0",
      "instanceLocation": "",
      "errors": [
        {
          "valid": false,
          "keywordLocation": "/oneOf/0/const",
          "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/0/const",
          "instanceLocation": "",
          "error": "Expected \"\\\"M\\\"\""
        }
      ]
    },
    {
      "valid": true,
      "keywordLocation": "/oneOf/1",
      "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/1",
      "instanceLocation": "",
      "annotations": [
        {
          "valid": true,
          "keywordLocation": "/oneOf/1/title",
          "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/1/title",
          "instanceLocation": "",
          "annotation": "Female"
        },
        {
          "valid": true,
          "keywordLocation": "/oneOf/1/title@mi-NZ",
          "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/1/title@mi-NZ",
          "instanceLocation": "",
          "annotation": "Wahine"
        }
      ]
    },
    {
      "valid": false,
      "keywordLocation": "/oneOf/2",
      "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/2",
      "instanceLocation": "",
      "errors": [
        {
          "valid": false,
          "keywordLocation": "/oneOf/2/const",
          "absoluteKeywordLocation": "https://json-everything.net/438509aca5#/oneOf/2/const",
          "instanceLocation": "",
          "error": "Expected \"\\\"D\\\"\""
        }
      ]
    }
  ]
}
```

You can see there's an entry with `"keywordLocation": "/oneOf/1/title@mi-NZ"` that has the annotation value you're looking for.

---

The new proposed format is a bit easier to read, but I don't think anyone but me supports it yet since it's not released.

```json
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/79e565d592#",
  "instanceLocation": "",
  "details": [
    {
      "valid": false,
      "evaluationPath": "/oneOf/0",
      "schemaLocation": "https://json-everything.net/79e565d592#/oneOf/0",
      "instanceLocation": "",
      "errors": {
        "const": "Expected \"\\\"M\\\"\""
      }
    },
    {
      "valid": true,
      "evaluationPath": "/oneOf/1",
      "schemaLocation": "https://json-everything.net/79e565d592#/oneOf/1",
      "instanceLocation": "",
      "annotations": {
        "title": "Female",
        "title@mi-NZ": "Wahine"
      }
    },
    {
      "valid": false,
      "evaluationPath": "/oneOf/2",
      "schemaLocation": "https://json-everything.net/79e565d592#/oneOf/2",
      "instanceLocation": "",
      "errors": {
        "const": "Expected \"\\\"D\\\"\""
      }
    }
  ]
}
```

**Comment by ThorFjelldalen** [comment]
Wonderful @gregsdennis! Thanks for the in-depth answer! 🙌
